### PR TITLE
fix(core): execute buffered requests during load API

### PR DESCRIPTION
**Pull request by MoumitaM**
## Description of the change
- Updated request buffering part during SDK initialization.
- Added a call to execute buffered requests at the end of load API.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#665]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/680)

### CODE REVIEW

File: analytics.js [hunk 1]
```diff
@@ -1265,12 +1276,20 @@ if (isValidArgsArray) {
...
   }
+
+  /**
+   * If defaultMethod is present in the buffered requests execute it.
+   */
+  if (argumentsArray.length > 0 && argumentsArray[0][0] === defaultMethod) {
+    argumentsArray[0].shift();
+    instance[defaultMethod](...argumentsArray[0]);
+    argumentsArray.shift();
+  }
```
Comment (saikumarrs) [inline analytics.js:1292]: Let's not execute the `load` API separately

File: analytics.js [hunk 2]
```diff
@@ -1185,15 +1194,17 @@ class Analytics {
...
+    analytics.toBeProcessedArray.forEach((x) => {
+      const event = [...x];
+      const method = event[0];
+      event.shift();
+      // logger.debug("=====from analytics array, calling method:: ", method)
+      analytics[method](...event);
+    });
+
+    analytics.toBeProcessedArray = [];
+  }
```
Comment (saikumarrs) [inline analytics.js:1207]: Would this work?

```suggestion
  // Process the queue only if the default API is its first entry 
  // or SDK is already loaded
  if (analytics.toBeProcessedArray.length && (analytics.loaded || analytics.toBeProcessedArray[0][0] === defaultMethod) ) {
  
    while (analytics.toBeProcessedArray.length > 0) {
      const event = [...analytics.toBeProcessedArray[0]];
      
      // remove the element from the queue
      analytics.toBeProcessedArray.shift();
      
      const method = event[0];
      event.shift();
      // logger.debug("=====from analytics array, calling method:: ", method)
      analytics[method](...event);
    }
  }
```

File: analytics.js [hunk 3]
```diff
@@ -1034,6 +1035,10 @@ class Analytics {
         } else {
           this.processResponse(200, res);
         }
+        /**
+         * In case load API is called later, execute the buffered requests.
+         */
```
Comment (saikumarrs) [inline analytics.js:1040]:
```suggestion
        // Execute any pending buffered requests
        // (needed if the load call was not previously buffered)
```

File: analytics.js [hunk 4]
```diff
@@ -1048,6 +1053,10 @@ class Analytics {
     } catch (error) {
       handleError(error);
     }
+    /**
+     * In case load API is called later, execute the buffered requests.
+     */
```
Comment (saikumarrs) [inline analytics.js:1058]:
```suggestion
    // Execute any pending buffered requests
    // (needed if the load call was not previously buffered)
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: analytics.js (hunk 1 above)
Comment (MoumitaM) [reply analytics.js]: Addressed

**Review by tech-goldenphoenix — CHANGES_REQUESTED** [review]

File: analytics.js [hunk 5]
```diff
@@ -1185,15 +1192,24 @@ class Analytics {
...
-    analytics[method](...event);
-  });
-
-  instance.toBeProcessedArray = [];
+  // Process the queue only if the default API is its first entry
+  // or SDK is already loaded
+  if (
+    analytics.toBeProcessedArray.length &&
+    (analytics.loaded || analytics.toBeProcessedArray[0][0] === defaultMethod)
+  ) {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js:1200]: @tech-goldenphoenix , Here `defaultMethod` is `load`. No other event will be processed till `analytics.loaded` is `true`. So, the `analytics.toBeProcessedArray` will still have `event a` and `event b`.

[the remark this answers is not in the document]
Comment (duannx) [reply analytics.js:1200]: @MoumitaM Let's say the state appears that: 
```js
analytics.loaded = false
analytics.toBeProcessedArray = [['load'], ['track', 'event a'], ['track', 'event b']]
```
The `if` statement will be `true`. So the `while` statement will be executed so these following methods will be fired in synchronous order: 
`['load'] => ['track', 'event a'] => ['track', 'event b']`

But if the `load` method has to load the polyfill, the method `loadAfterPolyfill` will be called asynchronously. And the property `analytics.loaded` will be set to true after all the events fired
`['load'] => ['track', 'event a'] => ['track', 'event b'] => 'load' method load the polyfill asynchronously => analytics.loaded = true`

Based on the above flow, `event a` and `event b` will be ignored because the `analytics.loaded` is not set to true yet

[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Apologies for the delay. Yes, you are right. Some events might be ignored depending on the polyfill and network strength. Refactored the code. 
Let us know if this works for you.

File: analytics.js [hunk 6]
```diff
@@ -1271,6 +1284,12 @@ if (isValidArgsArray) {
     }
     i += 1;
   }
+
+  if (instance.toBeProcessedArray.length && instance.toBeProcessedArray[0][0] === defaultMethod) {
+    instance.toBeProcessedArray[0].shift();
+    instance[defaultMethod](...instance.toBeProcessedArray[0]);
```
Comment (duannx) [inline analytics.js:1290]: If we execute the `load` methods here. All other methods will be ignored in the normal case when no polyfill is loaded. Because the `instance.toBeProcessedArray` is currently empty 
I think we should double check the test cases too because I see all the tests were passed

**Review by duannx — CHANGES_REQUESTED** [review]

File: analytics.js (hunk 6 above)
Comment (saikumarrs) [reply analytics.js:1290]: Yes. @MoumitaM, the polyfill-based load problem should be solved differently.
Let's revert this commit.

Comment (duannx) [reply analytics.js:1290]: Sorry. It's my bad. Actually, other methods won't be ignored because you make a manual call of `processDataInAnalyticsArray` after calling the `load` method. But it shows another error.
```js
// assume that instance.toBeProcessedArray = [['load', 'abc', 'xyz']]
instance.toBeProcessedArray[0].shift();
// now instance.toBeProcessedArray = [['abc', 'xyz']]
// executing the `load` method
instance[defaultMethod](...instance.toBeProcessedArray[0]);
// the function processDataInAnalyticsArray will be called inside the `load` method. Then it will throw an error since instance.toBeProcessedArray is equal to `[['abc', 'xyz']]`
// this line is executed after the function processDataInAnalyticsArray is called but it's too late
instance.toBeProcessedArray.shift();

```
Another thing is that the code flow is a bit complicated. The function `processDataInAnalyticsArray` is called at 3 different places. It's way hard to follow and maintain. My suggestion is: 
- Remove the `load` method from `instance.toBeProcessedArray` before executing it
- Push all other methods into `instance.toBeProcessedArray` before executing the `load` method
- Only call the `processDataInAnalyticsArray` function inside the `loadAfterPolyfill` function

This will ensure that all the methods will be executed after the `load` method regardless the polyfill is loaded or not.

Comment (MoumitaM) [reply analytics.js]: Addressed the concerns.

File: analytics.js [hunk 7]
```diff
@@ -1278,7 +1291,11 @@ parseQueryString(window.location.search);
 
 if (isValidArgsArray) argumentsArray.forEach((x) => instance.toBeProcessedArray.push(x));
 
-processDataInAnalyticsArray(instance);
+if (instance.toBeProcessedArray.length && instance.toBeProcessedArray[0][0] === defaultMethod) {
```
Comment (saikumarrs) [inline analytics.js:1294]: Add documentation here.

Comment (saikumarrs) [inline analytics.js:1294]: Extract the load event in the while block above and execute here directly.

Comment (MoumitaM) [reply analytics.js]: Addressed

Comment (MoumitaM) [reply analytics.js]: Addressed

File: analytics.js [hunk 8]
```diff
@@ -1185,15 +1192,21 @@ class Analytics {
...
-    const event = [...x];
-    const method = event[0];
-    event.shift();
-    // logger.debug("=====from analytics array, calling method:: ", method)
-    analytics[method](...event);
-  });
-
-  instance.toBeProcessedArray = [];
+  // Process the queue only if the default API is its first entry
+  // or SDK is already loaded
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply analytics.js]: Addressed

**Review by saikumarrs — APPROVED** [review]

**Review by duannx — APPROVED** [review]
LGTM!
